docs/architecture/adr-002-lite-client.md

```diff
@@ -0,0 +1,454 @@
+We also need to check what block this commit is for (ie. the hash of the
+header). 
+Since the commit is for a particular header, we can require all votes to be for that header, 
+and otherwise ignore them. Again, we use an associated type:
+
+```rust
+pub trait Commit {
+    type Vote: Vote;
+
+    fn header_hash(&self) -> Hash;
+    fn into_vec(&self) -> Vec<Option<Self::Vote>>;
+}
+```
+
+Note the `Option` here. When it is `None`, it indicates that either:
+
+- there was no vote from this validator
+- the validator voted nil
+- the validator voted for some other block
+
+We may want to be more strict here in distinguishing between these cases. 
+On the one hand, 
+[upcoming changes](https://github.com/tendermint/tendermint/blob/master/docs/architecture/adr-025-commit.md)
+to the commit structure will prevent votes from being included if they are from the
+wrong block.
+On the other, we may want the validator to verify the votes for nil - even
+though they don't contribute anything to the voting power, they serve as an
+extra validity check. I propose for now that we use the simple Option type and
+ignore the distinction between these cases, but that we revisit at a future
+date. 
+
+Note also that in Tendermint, commits are for a particular block ID, which
+includes both a header hash and a "parts set hash". The latter is completely
+irelevant to the light client, and can only be verified by downloading the full
+block. Hence it is effectively ignored here. It would be great if Tendermint
+could disentangle commits to the proposal block parts for gossip (ie. the parts
+set hash) from commits to the header itself (ie. the header hash), but that's
+left for the future.
+
+In the future, the Commit structure may change further to adopt an aggregate
+signature scheme. While we could pre-empt this by abstracting further, it's
+probably premature to do so now.
+
+### Implementing Traits
+
+The tendermint-rs library includes Header, Vote, Validator, ValidatorSet, and
+Commit data types. However, rather than use these types directly, the lite
+client library is written more abstractly to use traits that contain only the
+necessary information from these more concrete types. While this may turn out to
+be an unecessarily eager abstraction (as we do not forsee alternative
+implementations of these traits in the short term), it does provide a very clear
+depiction of what is required for lite client verification, and surfaces certain
+design issues in the underlying Tendermint blockchain (eg. the `BlockID` issue
+referenced above).
+
+This abstraction may also facilitate test generation, as we will not need to
+generate complete Tendermint data structures to test the lite client logic, only
+the elements it cares. While this provides a lot of flexibility in mocking out
+the types, we must be careful to ensure they match the semantics of the actual
+Tendermint types, and that we still test the verification logic sufficiently for
+the actual types.
```

Comment:
Clearly describing the rationale behind the decision to keep things abstract (and rather work with traits) 👍